R/react.R

```diff
@@ -0,0 +1,49 @@
+shiny_react_available <- function() {
+  requireNamespace("shiny.react", quietly = TRUE)
+}
+
+react_support <- function() {
+  if (shiny_react_available()) {
+    shiny::tagList(
+      shiny.react::reactDependency(),
+      shiny.react::shinyReactDependency(),
+      shiny::tags$script(shiny::HTML("
+        window.Rhino = {
+          registerReactComponents: (components) => {
+            window.jsmodule.RhinoReact ??= {};
+            Object.assign(window.jsmodule.RhinoReact, components);
+          },
+        };
+      "))
+    )
+  }
+}
+
+#' React components
+#'
+#' Access the React components defined in your app.
+#'
+#' @param name The name of the component.
+#' @return A function representing the component.
+#'
+#' @examples
+#' # Declare the component.
+#' TextBox <- react_component("TextBox")
+#'
+#' # Use the component.
+#' ui <- TextBox("Hello!", font_size = 20)
+#' @export
+react_component <- function(name) {
+  if (!shiny_react_available()) {
+    cli::cli_abort(
+      "To use React components in your app, please add {.pkg shiny.react} to dependencies."
+    )
+  }
```

Comment:
I have thought about it, but I couldn't conceive a good way to do it. Such a check would require two pieces of information:
1. What components are available (defined and registered with `Rhino.registerReactComponents()`).
2. What components are needed (declared with `react_component()` and used).

Here are some theoretical ideas to get the list of available components:
1. Parse the sources in `app/js` looking for calls to `Rhino.registerReactComponents()`.
2. Execute the built `app.min.js` and retrieve the `jsmodule.RhinoReact` object.
3. Register the components with some kind of a config file, which would be easier to parse then JS code.
4. Register the components with some kind of a mechanism, which would allow to generate a list of available components in easily-digestible form during `build_js()`.

And here are some theoretical ideas at which point we could determine what components are needed:
1. As an additional check for `lint_r()`.
2. When the `react_component()` function is called.
3. When the component function is called (the return value of `react_component()`).
4. When the application is launched.

I think some combination of (4) (available) and (2) (needed) could potentially have a chance to work, but it does seem like a serious endeavor.

I think the best we can do for now is to create an issue with these ideas :wink:

Reply:
Issue: #474.